File: src/LinkedList.Polynomial.Core/Polynomial.cs
using System.Text;

namespace LinkedList.Polynomial.Core;

/// <summary>
/// Polynomial represented as a singly linked list ordered by descending exponent.
/// </summary>
public sealed class Polynomial
{
    private sealed class PolynomialNode
    {
        public PolynomialNode(long coefficient, int exponent)
        {
            Coefficient = coefficient;
            Exponent = exponent;
        }

        public long Coefficient { get; set; }

        public int Exponent { get; }

        public PolynomialNode? Next { get; set; }
    }

    private PolynomialNode? _head;
    private PolynomialNode? _tail;

    public int TermCount { get; private set; }

    public bool IsZero => _head is null;

    public IEnumerable<(long Coefficient, int Exponent)> Terms
    {
        get
        {
            var current = _head;
            while (current is not null)
            {
                yield return (current.Coefficient, current.Exponent);
                current = current.Next;
            }
        }
    }

    public static Polynomial FromTerms(IEnumerable<(long Coefficient, int Exponent)> terms)
    {
        ArgumentNullException.ThrowIfNull(terms);

        var polynomial = new Polynomial();
        foreach (var (coefficient, exponent) in terms)
        {
            polynomial.AddTerm(coefficient, exponent);
        }

        return polynomial;
    }

    /// <summary>
    /// Inserts a term while preserving descending exponent order.
    /// Time: O(n), Space: O(1).
    /// </summary>
    public void AddTerm(long coefficient, int exponent)
    {
        if (exponent < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(exponent), "Exponent must be non-negative.");
        }

        if (coefficient == 0)
        {
            return;
        }

        if (_head is null)
        {
            AppendAsFirstTerm(coefficient, exponent);
            return;
        }

        if (exponent > _head.Exponent)
        {
            var newHead = new PolynomialNode(coefficient, exponent) { Next = _head };
            _head = newHead;
            TermCount++;
            return;
        }

        PolynomialNode? previous = null;
        var current = _head;

        while (current is not null && current.Exponent > exponent)
        {
            previous = current;
            current = current.Next;
        }

        if (current is not null && current.Exponent == exponent)
        {
            current.Coefficient += coefficient;

            if (current.Coefficient == 0)
            {
                RemoveNode(previous, current);
            }

            return;
        }

        var newNode = new PolynomialNode(coefficient, exponent) { Next = current };
        if (previous is null)
        {
            _head = newNode;
        }
        else
        {
            previous.Next = newNode;
        }

        if (current is null)
        {
            _tail = newNode;
        }

        TermCount++;
    }

    /// <summary>
    /// Adds two ordered linked-list polynomials using merge logic.
    /// Time: O(n + m), Space: O(n + m) for the result.
    /// </summary>
    public Polynomial Add(Polynomial other)
    {
        ArgumentNullException.ThrowIfNull(other);

        var result = new Polynomial();
        var left = _head;
        var right = other._head;

        while (left is not null || right is not null)
        {
            if (right is null || (left is not null && left.Exponent > right.Exponent))
            {
                result.AppendTermAssumeDescending(left!.Coefficient, left.Exponent);
                left = left.Next;
                continue;
            }

            if (left is null || right.Exponent > left.Exponent)
            {
                result.AppendTermAssumeDescending(right.Coefficient, right.Exponent);
                right = right.Next;
                continue;
            }

            var combinedCoefficient = left.Coefficient + right.Coefficient;
            if (combinedCoefficient != 0)
            {
                result.AppendTermAssumeDescending(combinedCoefficient, left.Exponent);
            }

            left = left.Next;
            right = right.Next;
        }

        return result;
    }

    /// <summary>
    /// Multiplies two polynomials.
    /// Time: O(n * m + k log k), where k is the number of unique exponents in the result.
    /// Space: O(k).
    /// </summary>
    public Polynomial Multiply(Polynomial other)
    {
        ArgumentNullException.ThrowIfNull(other);

        var result = new Polynomial();
        if (_head is null || other._head is null)
        {
            return result;
        }

        var accumulator = new Dictionary<int, long>();
        for (var left = _head; left is not null; left = left.Next)
        {
            for (var right = other._head; right is not null; right = right.Next)
            {
                var exponent = checked(left.Exponent + right.Exponent);
                var coefficient = checked(left.Coefficient * right.Coefficient);

                accumulator.TryGetValue(exponent, out var currentCoefficient);
                accumulator[exponent] = checked(currentCoefficient + coefficient);
            }
        }

        foreach (var entry in accumulator.OrderByDescending(item => item.Key))
        {
            if (entry.Value != 0)
            {
                result.AppendTermAssumeDescending(entry.Value, entry.Key);
            }
        }

        return result;
    }

    public override string ToString()
    {
        if (_head is null)
        {
            return "0";
        }

        var builder = new StringBuilder();
        var current = _head;

        while (current is not null)
        {
            var coefficient = current.Coefficient;
            if (coefficient != 0)
            {
                if (builder.Length == 0)
                {
                    if (coefficient < 0)
                    {
                        builder.Append("-");
                    }

                    AppendTerm(builder, Math.Abs(coefficient), current.Exponent);
                }
                else
                {
                    builder.Append(coefficient < 0 ? " - " : " + ");
                    AppendTerm(builder, Math.Abs(coefficient), current.Exponent);
                }
            }

            current = current.Next;
        }

        return builder.Length == 0 ? "0" : builder.ToString();
    }

    private void AppendAsFirstTerm(long coefficient, int exponent)
    {
        var node = new PolynomialNode(coefficient, exponent);
        _head = node;
        _tail = node;
        TermCount = 1;
    }

    private void AppendTermAssumeDescending(long coefficient, int exponent)
    {
        if (coefficient == 0)
        {
            return;
        }

        if (_head is null)
        {
            AppendAsFirstTerm(coefficient, exponent);
            return;
        }

        if (_tail is null || exponent >= _tail.Exponent)
        {
            throw new InvalidOperationException("Terms must be appended in strictly descending exponent order.");
        }

        var node = new PolynomialNode(coefficient, exponent);
        _tail.Next = node;
        _tail = node;
        TermCount++;
    }

    private void RemoveNode(PolynomialNode? previous, PolynomialNode current)
    {
        if (previous is null)
        {
            _head = current.Next;
        }
        else
        {
            previous.Next = current.Next;
        }

        if (current.Next is null)
        {
            _tail = previous;
        }

        if (_head is null)
        {
            _tail = null;
        }

        TermCount--;
    }

    private static void AppendTerm(StringBuilder builder, long absoluteCoefficient, int exponent)
    {
        if (exponent == 0)
        {
            builder.Append(absoluteCoefficient);
            return;
        }

        if (absoluteCoefficient != 1)
        {
            builder.Append(absoluteCoefficient);
        }

        builder.Append("x");
        if (exponent != 1)
        {
            builder.Append("^");
            builder.Append(exponent);
        }
    }
}

File: src/main/java/io/github/parkpawapon/linkedlist/polynomial/Polynomial.java
package io.github.parkpawapon.linkedlist.polynomial;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public final class Polynomial {
    private static final class Node {
        private long coefficient;
        private final int exponent;
        private Node next;

        private Node(long coefficient, int exponent) {
            this.coefficient = coefficient;
            this.exponent = exponent;
        }
    }

    private Node head;
    private Node tail;
    private int termCount;

    public static Polynomial fromTerms(Iterable<PolynomialTerm> terms) {
        Objects.requireNonNull(terms, "terms must not be null");

        Polynomial polynomial = new Polynomial();
        for (PolynomialTerm term : terms) {
            Objects.requireNonNull(term, "terms must not contain null elements");
            polynomial.addTerm(term.coefficient(), term.exponent());
        }

        return polynomial;
    }

    public int getTermCount() {
        return termCount;
    }

    public boolean isZero() {
        return head == null;
    }

    public List<PolynomialTerm> terms() {
        List<PolynomialTerm> result = new ArrayList<>(termCount);
        Node current = head;
        while (current != null) {
            result.add(new PolynomialTerm(current.coefficient, current.exponent));
            current = current.next;
        }

        return List.copyOf(result);
    }

    public void addTerm(long coefficient, int exponent) {
        validateExponent(exponent);
        if (coefficient == 0) {
            return;
        }

        if (head == null) {
            appendAsFirstTerm(coefficient, exponent);
            return;
        }

        if (exponent > head.exponent) {
            Node newHead = new Node(coefficient, exponent);
            newHead.next = head;
            head = newHead;
            termCount++;
            return;
        }

        Node previous = null;
        Node current = head;

        while (current != null && current.exponent > exponent) {
            previous = current;
            current = current.next;
        }

        if (current != null && current.exponent == exponent) {
            long updatedCoefficient = Math.addExact(current.coefficient, coefficient);
            if (updatedCoefficient == 0) {
                removeNode(previous, current);
            } else {
                current.coefficient = updatedCoefficient;
            }

            return;
        }

        Node newNode = new Node(coefficient, exponent);
        newNode.next = current;

        if (previous == null) {
            head = newNode;
        } else {
            previous.next = newNode;
        }

        if (current == null) {
            tail = newNode;
        }

        termCount++;
    }

    public Polynomial add(Polynomial other) {
        Objects.requireNonNull(other, "other must not be null");

        Polynomial result = new Polynomial();
        Node left = head;
        Node right = other.head;

        while (left != null || right != null) {
            if (right == null || (left != null && left.exponent > right.exponent)) {
                result.appendTermAssumeDescending(left.coefficient, left.exponent);
                left = left.next;
                continue;
            }

            if (left == null || right.exponent > left.exponent) {
                result.appendTermAssumeDescending(right.coefficient, right.exponent);
                right = right.next;
                continue;
            }

            long combinedCoefficient = Math.addExact(left.coefficient, right.coefficient);
            if (combinedCoefficient != 0) {
                result.appendTermAssumeDescending(combinedCoefficient, left.exponent);
            }

            left = left.next;
            right = right.next;
        }

        return result;
    }

    public Polynomial multiply(Polynomial other) {
        Objects.requireNonNull(other, "other must not be null");

        Polynomial result = new Polynomial();
        if (head == null || other.head == null) {
            return result;
        }

        Map<Integer, Long> accumulator = new HashMap<>();
        for (Node left = head; left != null; left = left.next) {
            for (Node right = other.head; right != null; right = right.next) {
                int exponent = Math.addExact(left.exponent, right.exponent);
                long coefficient = Math.multiplyExact(left.coefficient, right.coefficient);
                accumulator.merge(exponent, coefficient, Math::addExact);
            }
        }

        accumulator.entrySet().stream()
            .filter(entry -> entry.getValue() != 0)
            .sorted(Map.Entry.<Integer, Long>comparingByKey(Comparator.reverseOrder()))
            .forEach(entry -> result.appendTermAssumeDescending(entry.getValue(), entry.getKey()));

        return result;
    }

    @Override
    public String toString() {
        if (head == null) {
            return "0";
        }

        StringBuilder builder = new StringBuilder();
        Node current = head;

        while (current != null) {
            long coefficient = current.coefficient;
            if (coefficient != 0) {
                if (builder.length() == 0) {
                    if (coefficient < 0) {
                        builder.append("-");
                    }

                    appendTerm(builder, Math.abs(coefficient), current.exponent);
                } else {
                    builder.append(coefficient < 0 ? " - " : " + ");
                    appendTerm(builder, Math.abs(coefficient), current.exponent);
                }
            }

            current = current.next;
        }

        return builder.length() == 0 ? "0" : builder.toString();
    }

    private void appendAsFirstTerm(long coefficient, int exponent) {
        Node node = new Node(coefficient, exponent);
        head = node;
        tail = node;
        termCount = 1;
    }

    private void appendTermAssumeDescending(long coefficient, int exponent) {
        if (coefficient == 0) {
            return;
        }

        if (head == null) {
            appendAsFirstTerm(coefficient, exponent);
            return;
        }

        if (tail == null || exponent >= tail.exponent) {
            throw new IllegalStateException("Terms must be appended in strictly descending exponent order.");
        }

        Node node = new Node(coefficient, exponent);
        tail.next = node;
        tail = node;
        termCount++;
    }

    private void removeNode(Node previous, Node current) {
        if (previous == null) {
            head = current.next;
        } else {
            previous.next = current.next;
        }

        if (current.next == null) {
            tail = previous;
        }

        if (head == null) {
            tail = null;
        }

        termCount--;
    }

    private static void appendTerm(StringBuilder builder, long absoluteCoefficient, int exponent) {
        if (exponent == 0) {
            builder.append(absoluteCoefficient);
            return;
        }

        if (absoluteCoefficient != 1) {
            builder.append(absoluteCoefficient);
        }

        builder.append('x');
        if (exponent != 1) {
            builder.append('^').append(exponent);
        }
    }

    private static void validateExponent(int exponent) {
        if (exponent < 0) {
            throw new IllegalArgumentException("Exponent must be non-negative.");
        }
    }
}
